llama_index/indices/response/response_builder.py

```diff
@@ -576,6 +587,122 @@ def get_response(
             return stream_response
 
 
+class Accumulate(BaseResponseBuilder):
+    def __init__(
+        self,
+        service_context: ServiceContext,
+        text_qa_template: QuestionAnswerPrompt,
+        streaming: bool = False,
+        use_async: bool = False,
+    ) -> None:
+        super().__init__(service_context=service_context, streaming=streaming)
+        self.text_qa_template = text_qa_template
+        self._use_async = use_async
+
+    def flatten_list(self, md_array: List[List[Any]]) -> List[Any]:
+        return list(item for sublist in md_array for item in sublist)
+
+    def format_response(self, outputs: List[Any], separator: str) -> str:
+        responses: List[str] = []
+        for response, formatted_prompt in outputs:
+            self._log_prompt_and_response(
+                formatted_prompt, response, log_prefix="Initial"
+            )
+            responses.append(response or "Empty Response")
+
+        return separator.join(
+            [f"Response {index + 1}: {item}" for index, item in enumerate(responses)]
+        )
+
+    @llm_token_counter("aget_response")
+    async def aget_response(
+        self,
+        query_str: str,
+        text_chunks: Sequence[str],
+        separator: str = "\n---------------------\n",
+        **kwargs: Any,
+    ) -> RESPONSE_TEXT_TYPE:
+        """Apply the same prompt to text chunks and return async responses"""
+
+        tasks = self.get_response(query_str, text_chunks, **kwargs)
+
+        flattened_tasks = self.flatten_list(tasks)
+        outputs = await asyncio.gather(*flattened_tasks)
+
+        return self.format_response(outputs, separator)
+
+    @llm_token_counter("get_response")
+    def get_response(
+        self,
+        query_str: str,
+        text_chunks: Sequence[str],
+        separator: str = "\n---------------------\n",
+    ) -> str | List[List[Any]]:
+        """Apply the same prompt to text chunks and return responses"""
+
+        if self._streaming:
+            raise ValueError("Unable to stream in Accumulate response mode")
+
+        tasks = [
+            self._give_responses(
+                query_str,
+                text_chunk,
+            )
+            for text_chunk in text_chunks
+        ]
+
+        if self._use_async:
+            return tasks
```

Comment:
not super sure what is going on here.

Reply:
To help clarify, we use the flag `use_async` to enable async processing, even if the caller is not async (i.e. calling `get_response` with `use_async` turned on, make the behavior similar to `aget_response`).

Here, since the tasks would be a list of Coroutines, what we need to do here is actually run asyncio gather in an event loop. For this you can use `run_async_tasks` from `llama_index/async_utils.py`

Reply:
So I'm doing `if self._use_async:` here and returning the `tasks` so that `aget_response` can reuse the same function but handle the coroutines. Did you check out the unit test I added for async? My understanding was that `aget_response` would be called if `use_async` was turned on. Is that not the case? If not, would you mind pointing me to where the `use_async` param is passed?

Reply:
Ah I see, ya I can see why this is confusing. But basically, there are 3 cases:
1. caller calls `aget_response`: here we can use asyncio.gather directly
2. caller calls `get_response` with `use_async=False`: here the calls would be sequential and blocking
3. caller calls `get_response` with `use_async=True`: here we spin up an event loop, to run the LLM calls in async

Reply:
There's some historical reasons why we supported all 3 modes.

Reply:
Alllllllright. Thanks for that explanation, that clarifies a lot. I was very confused by why there seemed to be 2 variations of async calls. I definitely still don't understand the historical context, but that's ok. I had also seen the `run_async_tasks` function but was like "why would we need that for this aget function?". Take a look at the updated code and let me know what you think. I decided to separate `aget_response` and `get_response` to make it clearer.